clang/lib/Sema/SemaChecking.cpp

```diff
@@ -10619,6 +10619,37 @@ static std::optional<IntRange> TryGetExprRange(ASTContext &C, const Expr *E,
     case UO_AddrOf: // should be impossible
       return IntRange::forValueOfType(C, GetExprType(E));
 
+    case UO_Minus: {
+      if (E->getType()->isUnsignedIntegerType()) {
+        return TryGetExprRange(C, UO->getSubExpr(), MaxWidth, InConstantContext,
+                               Approximate);
+      }
+
+      std::optional<IntRange> SubRange = TryGetExprRange(
+          C, UO->getSubExpr(), MaxWidth, InConstantContext, Approximate);
+
+      if (!SubRange)
+        return std::nullopt;
+
+      return IntRange(SubRange->Width + 1, false);
```

Comment:
This looks correct to me but is subtle, so it might be worth adding a comment here explaining why this is correct. Maybe something like: If the range was previously non-negative, we need an extra bit for the sign bit. If the range was not non-negative, we need an extra bit because the negation of the most-negative value is one bit wider than that value.